### PR TITLE
Task/short circuit bounded range 1552

### LINKED ISSUE #1552: Short Circuit Visitors - Bounded Range
For any visitors that extend BaseVisitor, we have the option to short circuit at leaf nodes.

Improving the speed of our visitors will cut query planning time and, in some cases, speed up the time it takes to debug a query.

### PR DESCRIPTION
**Pull request by mineralntl**

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ShortCircuitBaseVisitor.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+import org.apache.commons.jexl2.parser.ASTSizeFunction;
+import org.apache.commons.jexl2.parser.ASTSizeMethod;
+import org.apache.commons.jexl2.parser.ASTStringLiteral;
+import org.apache.commons.jexl2.parser.ASTTernaryNode;
+import org.apache.commons.jexl2.parser.ASTTrueNode;
+import org.apache.commons.jexl2.parser.ASTUnaryMinusNode;
+import org.apache.commons.jexl2.parser.ASTWhileStatement;
+import org.apache.commons.jexl2.parser.SimpleNode;
+
+public class ShortCircuitBaseVisitor extends BaseVisitor {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ShortCircuitBaseVisitor.java:39]: Add a short comment explaining why these jexl nodes were chosen

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitor.java [hunk 2]
```diff
@@ -81,7 +85,32 @@ public Object visit(ASTNRNode node, Object data) {
         }
         
         return false;
-        
+    }
+    
+    // ensure we short circuit these nodes
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitor.java]: Add a brief comment explaining why these nodes were chosen, something like "short circuit on nodes that make up a bounded range"

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitor.java [hunk 3]
```diff
@@ -81,7 +85,32 @@ public Object visit(ASTNRNode node, Object data) {
...
+    public Object visit(ASTLENode node, Object data) {
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTGENode node, Object data) {
+        return data;
+    }
+    
+    @Override
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitor.java:113]: Add something like "Don't expect to see a bounded range inside of a function"

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitorTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+public class BoundedRangeDetectionVisitorTest {
+    private ShardQueryConfiguration config;
+    private MockMetadataHelper helper;
+    
+    @Before
+    public void init() throws Exception {
+        this.config = ShardQueryConfiguration.create();
+        Multimap<String,Type<?>> queryFieldsDatatypes = HashMultimap.create();
+        queryFieldsDatatypes.put("foo", new StringType());
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitorTest.java]: Uppercase fields.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitorTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        
+        this.helper = new MockMetadataHelper();
+        this.helper.setNonEventFields(queryFieldsDatatypes.keySet());
+    }
+    
+    @Test
+    public void testBoundedRange() throws ParseException {
+        String queryString = "((_Bounded_ = true) && (foo >= '1' && foo <= '10'))";
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(queryString);
+        assertTrue(BoundedRangeDetectionVisitor.mustExpandBoundedRange(config, helper, script));
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitorTest.java]: I'm always a fan of delegating to a test method so you can easily test multiple queries per method. 

At a minimum I would like to see two more tests added where the bounded range is the child of an OR and the child of an AND. For example `FOO == 'bar' && <bounded range>` and `FOO == 'bar' || <bounded range>`

**Review by apmoriarty — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitor.java (hunk 3 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitor.java:113]: Done

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitor.java (hunk 2 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitor.java]: Done

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitorTest.java (hunk 5 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitorTest.java]: Added as requested.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitorTest.java (hunk 4 above)
Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/BoundedRangeDetectionVisitorTest.java]: Done.

**Review by apmoriarty — APPROVED** [review]

**Review by avgAGB — APPROVED** [review]

**Pull request merged**